Add a tutorial about OpenJDK CRaC

pushkarnk opened the pull request:
This tutorial introduces CRaC and takes the reader through steps to build, checkpoint and restore a Spring Boot application.

rkratky commented on docs/tutorials/crac-use.md:
```diff
@@ -0,0 +1,122 @@
+(use-crac)=
+
+# Fast-startup for Spring Boot apps using OpenJDK CRaC 
```
```suggestion
# Fast start for Spring Boot applications using OpenJDK CRaC 
```

rkratky commented on docs/tutorials/crac-use.md:
```diff
@@ -0,0 +1,122 @@
+
+Tools like [`CRIU`](https://github.com/checkpoint-restore/criu), which stands for `checkpoint/restore in userspace`, support the dumping of the memory-snapshot of a running process, in any arbitrary state. Subsequently, the snapshot may be used to attempt a restore of the process in the same state. OpenJDK CRaC brings this capability to the Java runtime. 
+
+CRaC makes it possible to checkpoint a Java application, running in a state of 'peak performance'. Using the checkpoint snapshot, subsequent instances of the application could be spun up, almost instantaneously, in nearly* the same state of execution.
```
```suggestion
CRaC makes it possible to checkpoint a Java application, running in a state of 'peak performance'. Using the checkpoint snapshot, subsequent instances of the application could be spun up, almost instantaneously, in nearly the same state of execution.
```

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+apt update && apt install openjdk-21-crac-jdk-headless
+```
+
+## Checkpoint/Restore of Spring Boot applications
````
```suggestion
## Checkpointing and restoring Spring Boot applications
```

rkratky commented on docs/tutorials/crac-use.md:
```diff
@@ -0,0 +1,122 @@
+
+The following sections list steps to build, checkpoint and restore a Spring Boot application.
+
+## Build a Spring Boot application with CRaC support
```
```suggestion
### Building a Spring Boot application with CRaC support
```

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+```
+
+3. Point JAVA_HOME to the OpenJDK CRaC installation:
+```shell
+export JAVA_HOME=/usr/lib/jvm/java-21-openjdk-crac-amd64
+```
````
````suggestion

   ```none
   $ export JAVA_HOME=/usr/lib/jvm/java-21-openjdk-crac-amd64
   ```
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+export JAVA_HOME=/usr/lib/jvm/java-21-openjdk-crac-amd64
+```
+
+4. Build with maven:
+```shell
+mvn clean package
+```
````
````suggestion
4. Build with `maven`:

   ```none
   $ mvn clean package
   ```
````

rkratky commented on docs/tutorials/crac-use.md:
```diff
@@ -0,0 +1,122 @@
+
+If the build passes, a JAR file will be created under the `target` directory.
+
+## Checkpoint the Spring Boot application on start-up
```
```suggestion
### Checkpointing the Spring Boot application on start-up
```

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+## Checkpoint the Spring Boot application on start-up
+
+1. Start the application using this command:
+```shell
+$JAVA_HOME/bin/java -XX:CRaCCheckpointTo=$HOME/cr-image -jar target/spring-petclinic-3.5.0-SNAPSHOT.jar
+```
````
````suggestion

   ```none
   $ $JAVA_HOME/bin/java -XX:CRaCCheckpointTo=$HOME/cr-image \
     -jar target/spring-petclinic-3.5.0-SNAPSHOT.jar
   ```
```

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+$JAVA_HOME/bin/java -XX:CRaCCheckpointTo=$HOME/cr-image -jar target/spring-petclinic-3.5.0-SNAPSHOT.jar
+```
+
+The `-XX:CRaCCheckpointTo` option lets you configure the directory where the snapshot image will be dumped, on checkpoint.
````
```suggestion
The `-XX:CRaCCheckpointTo` option lets you configure the directory where the snapshot image is dumped, on checkpoint.
```

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+
+The `-XX:CRaCCheckpointTo` option lets you configure the directory where the snapshot image will be dumped, on checkpoint.
+
+The `PetClinic` application typically takes 5-7 seconds to start-up and become accessible:
+```{figure} /images/crac-use/01-normal-startup.png
+   :alt: Screenshot displaying a normal `PetClinic` application startup
+```
+
+The web-app should now be accessible at `http://localhost:8080`.
````
````suggestion
   The `PetClinic` application typically takes 5-7 seconds to start and become accessible:
   ```{figure} /images/crac-use/01-normal-startup.png
      :alt: Screenshot displaying a normal `PetClinic` application start
   ```

   The web app should now be accessible at `http://localhost:8080`.
```

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+
+
+2. Open another terminal and issue the `jcmd` command to list the running Java processes:
+```shell
+jcmd
+```
+This will produce output like:
+```
+4146 target/spring-petclinic-3.5.0-SNAPSHOT.jar
+4206 jdk.jcmd/sun.tools.jcmd.JCmd
+```
+Here, the `PetClinic` application is running in process with process-id 4146.
````
````suggestion

   ```none
   $ jcmd
   ```

   This produces output like:

   ```
   4146 target/spring-petclinic-3.5.0-SNAPSHOT.jar
   4206 jdk.jcmd/sun.tools.jcmd.JCmd
   ```

   Here, the `PetClinic` application is running in process with ID 4146.
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+Here, the `PetClinic` application is running in process with process-id 4146.
+
+3. To checkpoint, again use `jcmd` and issue this command:
+```shell
+jcmd 4146 JDK.checkpoint
+```
+The `PetClinic` application should now crash with a message reading "Killed".
+
+```{figure} /images/crac-use/02-app-killed.png
+   :alt: Screenshot displaying the `PetClinic` application killed on checkpoint.
+```
+
+The snapshot, which is a set of `.img` files, should be located in the configured directory. We used `$HOME/cr-data` as an example:
+
+```{figure} /images/crac-use/03-snapshot-dir.png
+   :alt: Screenshot showing the contents of the snapshot directory.
+```
````
````suggestion

   ```none
   $ jcmd 4146 JDK.checkpoint
   ```

   The `PetClinic` application should now crash with a message reading `Killed`.

   ```{figure} /images/crac-use/02-app-killed.png
      :alt: Screenshot displaying the `PetClinic` application killed on checkpoint.
   ```

   The snapshot, which is a set of `.img` files, should be located in the configured directory. We used `$HOME/cr-data` as an example:

   ```{figure} /images/crac-use/03-snapshot-dir.png
      :alt: Screenshot showing the contents of the snapshot directory.
   ```
````

rkratky commented on docs/tutorials/crac-use.md:
```diff
@@ -0,0 +1,122 @@
+
+## Restoring the Spring Boot application at checkpoint
+
+Using the snapshot produced at checkpoint, we attempt to restore the application in the same state.
```
```suggestion
Using the snapshot produced at checkpoint, restore the application in the same state.
```

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+Using the snapshot produced at checkpoint, we attempt to restore the application in the same state.
+
+Use this command to restore:
+```shell
+$JAVA_HOME/bin/java -XX:CRaCRestoreFrom=$HOME/cr-data
+```
````
````suggestion

```none
$ $JAVA_HOME/bin/java -XX:CRaCRestoreFrom=$HOME/cr-data
```
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,122 @@
+   :alt: Screen showing the application startup through CRaC restore.
+```
+
+## Practical considerations
````
```suggestion
## Practical considerations

```

rkratky commented on docs/tutorials/crac-use.md:
```diff
@@ -0,0 +1,122 @@
+
+## Practical considerations
+ - OpenJDK CRaC mostly benefits applications that are long-running and `stateful`.
+ - Ideally, the checkpoint process is carried out in a staging environment with load that is representative of the real-world load. The snapshot may then be used in production to rapidly spin-up application instances.
```
```suggestion
 - Ideally, the checkpoint process is carried out in a staging environment with a load that is representative of the real-world load. The snapshot may then be used in production to rapidly spin-up application instances.
```

rkratky reviewed (changes requested):
Thank you, @pushkarnk! I added a bunch of suggestions.

rkratky commented:
> CRaC is currently broken on Oracular and Plucky. The steps listed here will, for now, only work in a questing LXD container. I am unsure if this needs to be addressed in the tutorial.

Oops, overlooked this.

Hmm, there's no point in publishing docs about something that only works on an unreleased version. I mean, we can do it, but only with a big fat warning at the top of the page. And I wouldn't do even that -- it doesn't send a good message.

When you say "currently", does it mean it'll be fixed soon?

vpa1977 commented on docs/tutorials/crac-use.md:
```diff
@@ -0,0 +1,122 @@
+
+Java applications are characterized by a slow start-up, causing significant delays in application instances becoming available, which may be unacceptable in real-world deployments. Due to the inherent nature of the Java runtime, applications begin with a poor performance, go through a warm-up phase, eventually reaching a state of peak performance.
+
+Tools like [`CRIU`](https://github.com/checkpoint-restore/criu), which stands for `checkpoint/restore in userspace`, support the dumping of the memory-snapshot of a running process, in any arbitrary state. Subsequently, the snapshot may be used to attempt a restore of the process in the same state. OpenJDK CRaC brings this capability to the Java runtime. 
```
```suggestion
support dumping the memory snapshot of a running process
```

vpa1977 reviewed (commented):
Nice writeup! Google had some style suggestions, which I believe are essentially nits that improve clarity of the post.

pushkarnk commented:
> > CRaC is currently broken on Oracular and Plucky. The steps listed here will, for now, only work in a questing LXD container. I am unsure if this needs to be addressed in the tutorial.
> 
> Oops, overlooked this.
> 
> Hmm, there's no point in publishing docs about something that only works on an unreleased version. I mean, we can do it, but only with a big fat warning at the top of the page. And I wouldn't do even that -- it doesn't send a good message.
> 
> When you say "currently", does it mean it'll be fixed soon?

Moving this pull-request out of the draft state. OpenJDK CRaC now works on Ubuntu 25.04. 

Thanks for the reviews @rkratky and @vpa1977. I've applied all the suggested changes. I've also removed the images and replaced them with console text.

As for the use of `$` prompt, none of the commands that a user could copy-paste include it.

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,147 @@
+Here, the `PetClinic` application is running in process with process-id 4146.
+
+3. To checkpoint, again use `jcmd` and issue this command:
+```none
+jcmd 4146 JDK.checkpoint
+```
+The `PetClinic` application should now crash with a message reading "Killed".
+
+```
+...
+2025-09-04T06:53:56.670Z  INFO 7522 --- [Attach Listener] jdk.crac                                 : Starting checkpoint
+2025-09-04T06:53:56.677Z  INFO 7522 --- [Attach Listener] o.s.b.w.e.tomcat.GracefulShutdown        : Commencing graceful shutdown. Waiting for active requests to complete
+2025-09-04T06:53:56.695Z  INFO 7522 --- [tomcat-shutdown] o.s.b.w.e.tomcat.GracefulShutdown        : Graceful shutdown complete
+2025-09-04T06:53:56.700Z  WARN 7522 --- [Attach Listener] o.s.b.j.HikariCheckpointRestoreLifecycle : HikariDataSource (HikariPool-1) is not configured to allow pool suspension. This will cause problems when the application is checkpointed. Please configure allow-pool-suspension to fix this!
+2025-09-04T06:53:56.700Z  INFO 7522 --- [Attach Listener] o.s.b.j.HikariCheckpointRestoreLifecycle : Evicting Hikari connections
+Killed
+```
````
````suggestion

    ```none
    jcmd 4146 JDK.checkpoint
    ```
    The `PetClinic` application should now crash with a message reading "Killed".

    ```
    ...
    2025-09-04T06:53:56.670Z  INFO 7522 --- [Attach Listener] jdk.crac                                 : Starting checkpoint
    2025-09-04T06:53:56.677Z  INFO 7522 --- [Attach Listener] o.s.b.w.e.tomcat.GracefulShutdown        : Commencing graceful shutdown. Waiting for active requests to complete
    2025-09-04T06:53:56.695Z  INFO 7522 --- [tomcat-shutdown] o.s.b.w.e.tomcat.GracefulShutdown        : Graceful shutdown complete
    2025-09-04T06:53:56.700Z  WARN 7522 --- [Attach Listener] o.s.b.j.HikariCheckpointRestoreLifecycle : HikariDataSource (HikariPool-1) is not configured to allow pool suspension. This will cause problems when the application is checkpointed. Please configure allow-pool-suspension to fix this!
    2025-09-04T06:53:56.700Z  INFO 7522 --- [Attach Listener] o.s.b.j.HikariCheckpointRestoreLifecycle : Evicting Hikari connections
    Killed
    ```
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,147 @@
+Killed
+```
+
+The snapshot, which is a set of `.img` files, should be located in the configured directory. We used `$HOME/cr-data` as an example:
+
+```
+$ ls $HOME/cr-data
+core-3445.img  core-3455.img  core-3467.img  core-3479.img  core-3551.img  core-7531.img  core-7543.img  core-7555.img  core-7632.img  mm-3445.img
+core-3446.img  core-3456.img  core-3468.img  core-3480.img  core-7522.img  core-7532.img  core-7544.img  core-7556.img  core-7633.img  mm-7522.img
+core-3447.img  core-3457.img  core-3469.img  core-3481.img  core-7523.img  core-7533.img  core-7545.img  core-7557.img  dump4.log      pagemap-3445.img
+core-3448.img  core-3458.img  core-3472.img  core-3482.img  core-7524.img  core-7534.img  core-7546.img  core-7558.img  fdinfo-2.img   pagemap-7522.img
+core-3449.img  core-3459.img  core-3473.img  core-3483.img  core-7525.img  core-7535.img  core-7549.img  core-7559.img  files.img      pages-1.img
+core-3450.img  core-3460.img  core-3474.img  core-3545.img  core-7526.img  core-7536.img  core-7550.img  core-7560.img  fs-3445.img    pstree.img
+core-3451.img  core-3461.img  core-3475.img  core-3546.img  core-7527.img  core-7537.img  core-7551.img  core-7627.img  fs-7522.img    seccomp.img
+core-3452.img  core-3462.img  core-3476.img  core-3547.img  core-7528.img  core-7538.img  core-7552.img  core-7628.img  ids-3445.img   stats-dump
+core-3453.img  core-3463.img  core-3477.img  core-3549.img  core-7529.img  core-7539.img  core-7553.img  core-7629.img  ids-7522.img   timens-0.img
+core-3454.img  core-3466.img  core-3478.img  core-3550.img  core-7530.img  core-7540.img  core-7554.img  core-7631.img  inventory.img  tty-info.img
+```
````
````suggestion
    The snapshot, which is a set of `.img` files, should be located in the configured directory. We used `$HOME/cr-data` as an example:

    ```
    $ ls $HOME/cr-data
    core-3445.img  core-3455.img  core-3467.img  core-3479.img  core-3551.img  core-7531.img  core-7543.img  core-7555.img  core-7632.img  mm-3445.img
    core-3446.img  core-3456.img  core-3468.img  core-3480.img  core-7522.img  core-7532.img  core-7544.img  core-7556.img  core-7633.img  mm-7522.img
    core-3447.img  core-3457.img  core-3469.img  core-3481.img  core-7523.img  core-7533.img  core-7545.img  core-7557.img  dump4.log      pagemap-3445.img
    core-3448.img  core-3458.img  core-3472.img  core-3482.img  core-7524.img  core-7534.img  core-7546.img  core-7558.img  fdinfo-2.img   pagemap-7522.img
    core-3449.img  core-3459.img  core-3473.img  core-3483.img  core-7525.img  core-7535.img  core-7549.img  core-7559.img  files.img      pages-1.img
    core-3450.img  core-3460.img  core-3474.img  core-3545.img  core-7526.img  core-7536.img  core-7550.img  core-7560.img  fs-3445.img    pstree.img
    core-3451.img  core-3461.img  core-3475.img  core-3546.img  core-7527.img  core-7537.img  core-7551.img  core-7627.img  fs-7522.img    seccomp.img
    core-3452.img  core-3462.img  core-3476.img  core-3547.img  core-7528.img  core-7538.img  core-7552.img  core-7628.img  ids-3445.img   stats-dump
    core-3453.img  core-3463.img  core-3477.img  core-3549.img  core-7529.img  core-7539.img  core-7553.img  core-7629.img  ids-7522.img   timens-0.img
    core-3454.img  core-3466.img  core-3478.img  core-3550.img  core-7530.img  core-7540.img  core-7554.img  core-7631.img  inventory.img  tty-info.img
    ```
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,147 @@
+Using the snapshot produced at checkpoint, restore the application in the same state.
+
+Use this command to restore:
+```none
+$JAVA_HOME/bin/java -XX:CRaCRestoreFrom=$HOME/cr-data
+```
````
````suggestion

```none
$JAVA_HOME/bin/java -XX:CRaCRestoreFrom=$HOME/cr-data
```
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,147 @@
+## Building a Spring Boot application with CRaC support
+
+1. Clone the `Spring Boot PetClinic` application:
+```none
+git clone https://github.com/spring-projects/spring-petclinic && \
+cd spring-petclinic
+```
````
````suggestion

    ```none
    git clone https://github.com/spring-projects/spring-petclinic && \
    cd spring-petclinic
    ```
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,147 @@
+
+
+2. Open another terminal and issue the `jcmd` command to list the running Java processes:
+```none
+jcmd
+```
+This produces output like:
+```
+4146 target/spring-petclinic-3.5.0-SNAPSHOT.jar
+4206 jdk.jcmd/sun.tools.jcmd.JCmd
+```
+Here, the `PetClinic` application is running in process with process-id 4146.
````
````suggestion

    ```none
    jcmd
    ```

    This produces output like:

    ```
    4146 target/spring-petclinic-3.5.0-SNAPSHOT.jar
    4206 jdk.jcmd/sun.tools.jcmd.JCmd
    ```

    Here, the `PetClinic` application is running in process with process-id 4146.
````

rkratky reviewed (changes requested):
Thank you, @pushkarnk. I added some formatting fixes.

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,158 @@
+    The `-XX:CRaCCheckpointTo` option lets you configure the directory where the snapshot image is saved, on checkpoint.
+
+    The PetClinic application typically becomes available for requests within 5-7 seconds of startup.
+    ```
````
````suggestion

    ```none
````

rkratky commented on docs/tutorials/crac-use.md:
````diff
@@ -0,0 +1,158 @@
+
+    This produces output like:
+
+    ```
````
````suggestion
    ```none
````

rkratky reviewed (approved):
I just realized I missed two little things. I added the suggestions but pre-approving for merging.

rkratky reviewed (approved):
Thanks!
